### ISSUE TITLE
Payroll US Api missing PayItems endpoint in AmericanPayroll.cs

### ISSUE THREAD

**Issue by Parjinder** [XeroAPI/Xero-Net#286]
I was using Xero US Payroll api.But i was having issues with PayItems when I try to Find EarningTypeID based on another field.

It is not using Api.PayItems,Where().Find()

like in Australia Api i.e. Api.PayItems,Where().Find()

**Comment by Parjinder** [comment]
Is there any updates regarding my query?

**Comment by MJMortimer** [comment]
Hey @Parjinder.

Unfortunately the Payroll API does not provide any functionality for the Where clause filter. This SDK, because of how it has been built, will allow you to provide a query for the Where clause, but it will have no effect. 

I'll have to look into ways of providing feedback at run time when trying to use a Where clause filter on endpoints that don't support it, and I'll also look into adding a note in the Readme.

**Comment by Parjinder** [comment]
Hi @MJMortimer 

Thanks for the update.

So is there any alternative from where i can get EarningsTypeID in order to add salary and wages types for individual employees using Api (US)

**Employee Entpoints**
<Employees>
  <Employee>    
    <SalaryAndWages>
      <SalaryAndWage>
        <EarningsTypeID>5a467a52-e60e-4a3d-bca8-3447a8e1fd79</EarningsTypeID>
        <SalaryWagesType>HOURLY</SalaryWagesType>
        <HourlyRate>40.000000</HourlyRate>
        <StandardHoursPerWeek>40.0000</StandardHoursPerWeek>
      </SalaryAndWage>
    </SalaryAndWages>
</Employees>
</Employee>

**Comment by thevalyreangroup** [comment]
@Parjinder @MJMortimer , i too just ran into a similar issue today.  I had to build my own fork of the library and add the endpoint and methods needed to the American Payroll class.  In fact, AmericanPayroll did not have any of the endpoints needed to get PayItems, only the models.  I have not tested the WHERE filter because i did not need that for my implementation, but i don't see why it shouldn't work with what i've implemented as i modeled the methods after the Employees endpoint.  But at least now i can get a list of all pay items. 
![screen shot 2017-07-20 at 3 38 55 pm](https://user-images.githubusercontent.com/22800795/28435765-64ba56f4-6d62-11e7-863c-416bd8dac332.png)

**Comment by Parjinder** [comment]
Hi @thevalyreangroup 

Can you please share the information/code about how you get the list of pay items?
I also tried by creating methods but no luck.

Thanks

**Comment by thevalyreangroup** [comment]
@parjinder Sure.  I can update my fork on Github with my changes and send you the Repo link tomorrow morning.

**Comment by thevalyreangroup** [comment]
@MJMortimer, I'll also create a PR with the changes as well.

**Comment by thevalyreangroup** [comment]
@Parjinder Here's a working fork with the PayItems endpoint. https://github.com/thevalyreangroup/Xero-Net

Just get a new instance of the payroll api, then call `payrollApi.PayItems.Find().ToList();`

**Comment by Parjinder** [comment]
Hi @thevalyreangroup 

Thanks for the update.

Just one question, was the updates also available via nuget package for xero sdk?

Thanks

**Comment by thevalyreangroup** [comment]
@parjinder, no.  You will need to download and compile this yourself.  It would only be a nuget package once the PR is merged and updated to Nuget.

**Comment by Parjinder** [comment]
Hi @thevalyreangroup 

Some issues i faced when working with Employee Endpoint 

1) I got the following error when try to add worklocation and OpeningBalances for Employee using American api  Employee endpoint
{"Type definitions should start with a '{', expecting serialized type 'ApiException', got string starting with: <Response xmlns:i=\"http://www.w3.org/2001/XMLSchem"}

2) When i try to add salaryandwages for employee it shows the validation exception. for eg. Invalid SalaryandWageID .Although Employee Endpoint POST method does not use SalaryandWageID.

https://developer.xero.com/documentation/payroll-api-us/employees#POST

3) There are no option for taxes in Employee Endpoint api.

**Comment by thevalyreangroup** [comment]
Hey @parjinder, I am not a xero employee nor am I an active contributor to this SDK.  I've just found some bugs and missing things with my use of it.

I do t have any comments on #1 or #3 but for 
2) This I have experienced and will be my next fix.  Although, I do not have that problem when adding or updating an employee if you omit the SalaryAndWage.  Only when trying to specifically add that I get this issue.  I'll let you know when I've figured it out or if the API team fixes it first.  You should probably open a separate issue for that as well.

Were you able to use my fork for your payroll items?

**Comment by Parjinder** [comment]
Hi @thevalyreangroup,I was able to get list of payitems.thank you for your help.

**Comment by Parjinder** [comment]
Hi @thevalyreangroup 

2nd point- SalaryandWage

If I commented the SalaryAndWagesID,then i was able to add salary and wages for individual employees

namespace Xero.Api.Payroll.America.Model
{
    [DataContract(Namespace = "")]
    public class SalaryAndWage
    {
       // [DataMember(Name = "SalaryAndWagesID")]
       // public Guid Id { get; set; }

        [DataMember(Name = "EarningsTypeID")]
        public Guid EarningsTypeId { get; set; }

        [DataMember]
        public SalaryWagesType SalaryWagesType { get; set; }

        [DataMember]
        public decimal HourlyRate { get; set; }

        [DataMember]
        public decimal AnnualSalary { get; set; }

        [DataMember]
        public decimal StandardHoursPerWeek { get; set; }

        [DataMember]
        public DateTime EffectiveDate { get; set; }
    }
}

**Comment by thevalyreangroup** [comment]
@pqrjinder, nice!  I will have to do that. Have you tried reading the salary and wage object of the employee when retrieving? I always found the object to be NULL even though a Salary and wage object exists in xero.

**Comment by Parjinder** [comment]
Hi @thevalyreangroup 

Yes.I also got NULL value when retrieving the Employee.

**Comment by MJMortimer** [comment]
Hey @Parjinder, for this issue
`When i try to add salaryandwages for employee it shows the validation exception. for eg. Invalid SalaryandWageID .Although Employee Endpoint POST method does not use SalaryandWageID.`
I suspect the issue is the that when serializing the SalaryAndWage object it will serialize the SalaryandWageID to the default empty guid when not set. Can you please try changing the definition to the following?

```
[DataMember(Name = "SalaryAndWagesID", EmitDefaultValue=false)]  
public Guid Id { get; set; }
```

**Comment by thevalyreangroup** [comment]
@MJMortimer, so are you saying there is no way to update the SalaryAndWages for an employee?

**Comment by MJMortimer** [comment]
Nah, what I mean is that when trying to create the employee, when you add a new SalaryAndWages, it will serialize out an empty guid for the SalaryAndWagesID, which I suspect is causing the validation error about invalid ID. By setting EmitDefaultValue = false, it will mean that when not set it won't serialize and send the empty guide to the API

**Comment by MJMortimer** [comment]
The PayItems endpoint and the fix for poor serialization of the SalrayAndWagesID property has be added in @thevalyreangroup PR here #287 

@Parjinder was there anything else outstanding in this issue?

**Comment by MJMortimer** [comment]
Closing this issue, if anything else is still outstanding it can be opened in a fresh issue
